filters/diag/histogram.go

```diff
@@ -0,0 +1,123 @@
+package diag
+
+import (
+	"math/rand"
+	"time"
+
+	"github.com/zalando/skipper/filters"
+)
+
+type (
+	histSpec struct {
+		typ string
+	}
+
+	histFilter struct {
+		response bool
+		sleep    func(time.Duration)
+
+		weights []float64
+		sum     float64
+		buckets []time.Duration
+	}
+)
+
+// NewHistogramRequestLatency creates filters that add latency to requests according to the histogram distribution.
+// It expects a list of number and duration string pairs that defines a histogram.
+// Duration strings define consecutive bucket boundaries starting from implicit zero and numbers define bucket weights.
+// The filter randomly selects a bucket according to its weight and then sleeps for a random duration in between bucket boundaries.
+// The probability of selecting a bucket equals to its weight divided by the sum of all bucket weights.
+// Eskip example:
+//
+//	r: * -> histogramRequestLatency(50, "5ms", 0, "10ms", 30, "15ms", 20, "20ms") -> "https://www.example.org";
+//
+// The example above adds a latency
+// * between implicit 0ms and 5ms to 50% of the requests
+// * between 5ms and 10ms to 0% of the requests
+// * between 10ms and 15ms to 30% of the requests
+// * and between 15ms and 20ms to 20% of the requests.
+func NewHistogramRequestLatency() filters.Spec {
+	return &histSpec{typ: filters.HistogramRequestLatencyName}
+}
+
+// NewHistogramResponseLatency creates filters that add latency to responses according to the histogram distribution, similar to NewHistogramRequestLatency.
+func NewHistogramResponseLatency() filters.Spec {
+	return &histSpec{typ: filters.HistogramResponseLatencyName}
+}
+
+func (s *histSpec) Name() string {
+	return s.typ
+}
+
+func (s *histSpec) CreateFilter(args []any) (filters.Filter, error) {
+	if len(args) == 0 || len(args)%2 != 0 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	f := &histFilter{
+		response: s.typ == filters.HistogramResponseLatencyName,
+		sleep:    time.Sleep,
+		weights:  make([]float64, len(args)/2),
+		buckets:  make([]time.Duration, len(args)/2),
+	}
+
+	for i := 0; i < len(args); i += 2 {
+		weight, ok := args[i].(float64)
+		if !ok || weight < 0 {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		f.weights[i/2] = weight
+		f.sum += weight
+
+		ds, ok := args[i+1].(string)
+		if !ok {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		max, err := time.ParseDuration(ds)
+		if err != nil || max <= 0 {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		if i > 0 && max <= f.buckets[i/2-1] {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		f.buckets[i/2] = max
+	}
+
+	if f.sum > 0 {
```

Comment:
shouldn't this return an error instead of going around it?

Reply:
We use relative weights in other cases (f.e. `Traffic()`), so should be fine

Reply:
When sum is zero filter becomes a noop. Maybe we should disallow all zero weights and indeed return an error here.

Reply:
Fixed